### DISCUSSION TITLE
Private Key specification in strongswan version 4.5.2

### DISCUSSION THREAD

**Discussion by brobbinsharris** [strongswan/strongswan#1297] [Q&A]
Hi,
I am currently testing with 2 versions of strongswan - version 5.8.1 and version 4.5.2. I have generated my certificates and private keys and the connection works well with version 5.8.1 - where we have both pkcs1 and pkcs8 plugins loaded. Here is the script I use to generate my certificates and keys. In the case that is failing strongswan 4.5.2 is the responder.

#Generate a Self-Signed CA Certificate as root CA
openssl req \
       -newkey rsa:2048 -nodes -keyout result/ca_key.pem \
       -config openssl_ca.cnf -extensions v3_ca \
       -x509 -days 3650 \
       -out result/ca_cert.pem

#Generate certificate signing request (CSR) for initiator certificate
openssl req \
                -new -newkey rsa:2048 -nodes \
                -config openssl_i.cnf -extensions usr_cert \
                -keyout result/initiator_key.pem \
                -out result/initiator_req.pem

#Signing the initiator certificate with the root CA
openssl x509 \
                -req -in result/initiator_req.pem \
                -sha256 -days 3650 -extfile openssl_i.cnf -extensions usr_cert \
                -CA result/ca_cert.pem -CAkey result/ca_key.pem -CAcreateserial \
                -out result/initiator_cert.pem

#Generate certificate signing request (CSR) for responder certificate
openssl req \
                -new -newkey rsa:2048 -nodes \
                -config openssl_r.cnf -extensions usr_cert \
                -keyout result/responder_key.pem \
                -out result/responder_req.pem

#Signing the responder certificate with the root CA
openssl x509 \
                -req -in result/responder_req.pem \
                -sha256 -days 3650 -extfile openssl_r.cnf -extensions usr_cert \
                -CA result/ca_cert.pem -CAkey result/ca_key.pem -CAcreateserial \
                -out result/responder_cert.pem

---------------
When i run this set of certificates with strongswan version 4.5.2 i get an error that my responder pubkey does not have a private key. 
I then see in the log that I have an error when opening up the responder_key.pem that is in format for pkcs8.

openssl asn1parse -in responder_key.pem
    0:d=0  hl=4 l=1213 cons: SEQUENCE
    4:d=1  hl=2 l=   1 prim: INTEGER           :00
    7:d=1  hl=2 l=  13 cons: SEQUENCE
    9:d=2  hl=2 l=   9 prim: OBJECT            :rsaEncryption
   20:d=2  hl=2 l=   0 prim: NULL
   22:d=1  hl=4 l=1191 prim: OCTET STRING      [HEX DUMP]:308204A30201000282010100E11B435658E7BFC945201258A0B9A80090BA8DACF7E945F637C838369D4A41FA1EE674DC6266A8A8FA609CB26090E376BC39B9E0F16B0FB39BF537C46D784C355E986159D145EF726EA1C754FB44C517876B24E7258395CBE4E2AF2C2D970B877C5274A0BBEFB6BF95C377A98659A848D0E08F21A5D20F3B031CEDD7A2E171DADC65E226BB3C620AF8A021D7BE19D2057EC096EC8ECF67ACF1F70552E5AD12F8C2A0C445B1E5CE216A222EA145ADF36D7565465E792CBCE1D354E7015A54A5BFB167D8D572F2340E545A4A7C5ED51A53CDE2BC9AD2746AF501EBA576E5BA20B036017EFF05696C1219880C505B1F127EE0737A0AC02A4AEAAEE4626FBB1E96CB020301000102820100311B7A7826D2FFD2CFA2AC7DEB2A354D5245704CD094722FC1818B49D6CC715238557C1EEBD48B6CB99A43F4E1FE19B13864BB7928117C0E69B2782500E0A2776E187DFC531C179D762712CE26369D6C819D4D22A3DB6DDD5A245DA49C997BE965A4B34EB1511D23F46EFCF6DD99478E030B4A0E44176C3D2A73B02B3F9C40F2DE2E9E3D2F62261CB25E911F3141134F5338440F3F3F133BED35E4C12D964DC4502F8DDFC20727CE1DB64DAA1CE4F96F770E6F9EDDC7A4146564E4E4AB12093C1F9E52210857324D4B902600BDD07F29EB6625E7BB27DA58F4C1727B86960D2BF0AFADC8CE7D00760BA99857993A38D29A69872690EB3C8546B2A6BDDD40F0D902818100F03903ED6A240B5244276B206A175CA8B054BEBE1F5BD1A490A4C93FAA55D1B742CECBBA0DA7742A0114DF99A27A17CE503D6D356EC8CC1E369A332F55E1DE134EC12492DC1C89146F8666A50886FE297BF453FEB4413FF6A7B6224A8BEE5B0BAAFE8AF651B2E30A3EFA58BA942DDA4E67FF4CC30030E27E2EF74BFF2F0BBF7F02818100EFE417544597F1E57A9858369D38CAE8D6AA501F9F2A3303698B11878F5D48683B02A7E454BE3CEFAA6FC8BF12AECFBE018AED5A4ADC3C0A271E5350CF339EDCF2DB633F6954EE6204D14894DB2060FCEB2B1AA1CB04480A4FC7D5B3AA00FE199D315B579289730170F3839C9FF6E50E3AF677BB2B77750E61F3F939D848CEB502818015FE224DEA19C2AF956DC671014582462F5E92D7F52702DA7E2398092C1F0F4427E9A6DD31C2D2EFC58993AF0D21006CF30BACA3E7B09A7F69A0546F50FD8CC757BE770ABAFC21D0AD2530E08B1750CD9B7B92F50EABAA9F7A5B426870C731AEAFC8FE0FEF0AE19E5E85ACF9E7F70AB582EFECA72DAD4C0F215DA2A78A371921028181008B2B03EDBBC1FE6F66C3E2C92C44BA5281D6A44CF405833AC38C0E4B5A7C7F791862DE0BB4ADF8A40178803979193923BDB95F2136B4B5901E49958417F096850FCEFEB0D6E692B5029E7741042EB09589C8A3D6A0F486E6948A160BF8E04289918C04B5B5295032D1D833CE356F024B5321912C607718CD933DFAA3F839A4D10281806F603D1632D012B9E04712262F4505CAA9C5132ADCE6E3439C51DE47D68CE5D41D5AA96DFBBDF9C560053A48E455537F73E90FD893FA624E99ED45FA083100781E3026FC79AED39E8D5F0FD3BAE7121CE9D95235747B829FCCD62D2917BD42AD1F72396EEF63912244511D248DD899F61A8F3CA7F52D05AEA59D80D04CEAF513

My strongswan version 4.5.2 does not have pkcs8 plugin (i only have pkcs1) and when reading the asn1 string it fails when reading 0x30 in the fix Hex location when it expects to see 0x02.

I tried converting the responder_key.pem from pkcs8 to pkcs1 but this did not work.

how can I generate my private key to be compatible with asn1 and work with pkcs1?

I also tried "openssl rsa" to generate a pkcs1 format but this did not produce the 0x02 identifier.

thanks

**Comment by strongX509** [comment]
You could generate the private key with the strongSwan pki tool
````
pki --gen --type rsa --size 2048 --outform pem > key.pem
````
which encodes the key in PKCS#1 format.

BTW - strongSwan 4.5.2 is so old that you really shouldn't use it anymore.

**Reply by brobbinsharris** [reply]
Thank you - I will give that a try - agree with 4.5.2 be so old - but unfortunately we are stuck with it and cannot upgrade it.

**Reply by brobbinsharris** [reply]
Hi - I am still having an issue after using the PKI tool.

My Linux strongswan is the responder in this simple point to point scenario. In Version 5.8.1 after i start strongswan - here is my certificate.
  subject:  "C=US, ST=NJ, L=Clifton, O=L3, CN=Responder"
  issuer:   "C=US, ST=NJ, L=Clifton, O=L3, OU=EW, CN=Test_CA"
  validity:  not before Sep 08 11:37:31 2022, ok
             not after  Sep 05 11:37:31 2032, ok (expires in 3288 days)
  serial:    a0:b6:78:ac:12:c3:7a:59
  flags:
  authkeyId: 67:4f:b6:b0:41:31:89:a9:f4:dc:0e:e2:db:6a:7d:63:2f:c7:83:46
  subjkeyId: 52:f7:83:7e:a0:79:20:b8:86:7c:ba:c6:a4:04:a1:70:1c:8a:cc:e0
  pubkey:    RSA 2048 bits, has private key
  keyid:     46:ef:40:dd:bf:fe:d8:23:6c:cf:90:07:57:c1:64:b3:41:0e:dd:20
  subjkey:   52:f7:83:7e:a0:79:20:b8:86:7c:ba:c6:a4:04:a1:70:1c:8a:cc:e0

The pubkey of RSA 2048 needs to have a private key - which is the responder_key.pem.
root@xilinx-zcu102-20202:/etc/ipsec.d/private# ls -l
total 4
-rw-r--r--    1 root     root          1675 Sep  2 00:31 responder_key.pem
root@xilinx-zcu102-20202:/etc/ipsec.d/private#
root@xilinx-zcu102-20202:/etc# cat ipsec.secrets
# ipsec.secrets - strongSwan IPsec secrets file
: RSA responder_key.pem
root@xilinx-zcu102-20202:/etc#

When i run the PKI tool to generate a new responder_key.pem private key the cert looks like this.
  subject:  "C=US, ST=NJ, L=Clifton, O=L3, CN=Responder"
  issuer:   "C=US, ST=NJ, L=Clifton, O=L3, OU=EW, CN=Test_CA"
  validity:  not before Sep 08 11:37:31 2022, ok
             not after  Sep 05 11:37:31 2032, ok (expires in 3291 days)
  serial:    a0:b6:78:ac:12:c3:7a:59
  flags:
  authkeyId: 67:4f:b6:b0:41:31:89:a9:f4:dc:0e:e2:db:6a:7d:63:2f:c7:83:46
  subjkeyId: 52:f7:83:7e:a0:79:20:b8:86:7c:ba:c6:a4:04:a1:70:1c:8a:cc:e0
  pubkey:    RSA 2048 bits
  keyid:     46:ef:40:dd:bf:fe:d8:23:6c:cf:90:07:57:c1:64:b3:41:0e:dd:20
  subjkey:   52:f7:83:7e:a0:79:20:b8:86:7c:ba:c6:a4:04:a1:70:1c:8a:cc:e0

This is not correct and when trying to bring up the connection with the Initiator i get this error.
Sep  2 00:11:25 xilinx-zcu102-20202 daemon.info charon: 15[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
Sep  2 00:11:25 xilinx-zcu102-20202 daemon.info charon: 15[IKE] no private key found for 'C=US, ST=NJ, L=Clifton, O=L3, CN=Responder'
Sep  2 00:11:25 xilinx-zcu102-20202 daemon.info charon: 15[ENC] generating IKE_AUTH response 1 [ N(AUTH_FAILED) ]
Sep  2 00:11:25 xilinx-zcu102-20202 daemon.info charon: 15[NET] sending packet: from 192.168.0.10[500] to 192.168.0.11[500] (80 bytes)
Sep  2 00:11:25 xilinx-zcu102-20202 daemon.info charon: 15[IKE] IKE_SA ikev2-vpn[3] state change: CONNECTING => DESTROYING

Any ideas?

thanks

**Reply by brobbinsharris** [reply]
I believe my issue is that after creating the new RSA Key in pkcs1 format - i did not generate a new "certificate signing request" for the response certificate and did not sign the new response certificate.

Any help in this areas is appreciated to create this new key, and response certificate etc with the pki tool.

thanks

**Reply by tobiasbrunner** [reply]
There is a [tutorial for pki](https://docs.strongswan.org/docs/5.9/pki/pkiQuickstart.html) in the docs.

**Reply by brobbinsharris** [reply]
thanks for the quick feedback will take a look

**Reply by brobbinsharris** [reply]
the main point of my trouble is in the example in the tutorial.

pki --print --in moonCert.pem
subject:  "C=CH, O=strongSwan, CN=moon.strongswan.org"
issuer:   "C=CH, O=strongSwan, CN=strongSwan Root CA"
validity:  not before May 19 10:28:19 2017, ok
           not after  May 19 10:28:19 2022, ok (expires in 1825 days)
serial:    01
altNames:  moon.strongswan.org
flags:     serverAuth
CRL URIs:  http://crl.strongswan.org/strongswan.crl
authkeyId: 2b:95:14:5b:c3:22:87:de:d1:42:91:88:63:b3:d5:c1:92:7a:0f:5d
subjkeyId: 60:9d:de:30:a6:ca:b9:8e:87:bb:33:23:61:19:18:b8:c4:7e:23:8f
pubkey:    ED25519 256 bits
keyid:     39:1b:b3:c2:34:72:1a:01:08:40:ce:97:75:b8:be:ce:24:30:26:29
subjkey:   60:9d:de:30:a6:ca:b9:8e:87:bb:33:23:61:19:18:b8:c4:7e:23:8f

need the RSA pubkey to have a recognize a private key.

thanks

**Reply by tobiasbrunner** [reply]
`pki` does never print that information. How could it if you only print/load a certificate? The `pubkey: ..., has private key` output is only ever printed in `swanctl --list-certs` if the daemon also has the private key loaded for a specific certificate.

**Reply by brobbinsharris** [reply]
sorry - yes agree - you will see this after running and issuing "ipsec listcerts" - at least for the older versions i am using. If it doesn't show up there then there is a problem.

thanks for the help

**Reply by brobbinsharris** [reply]
with only pkcs1 plugin installed from charon - I run these 2 commands

pki --gen --type rsa --outform pem > responder_key.pem

pki --pub --in responder_key.pem | pki --issue --cacert ca_cert.pem --cakey ca_key.pem --dn "..." --outform pem --san peer > responder_cert.pem

The pki --issue command gives me the error.

ASN1 tag 0x02 expected, but is 0x30

**Reply by tobiasbrunner** [reply]
> with only pkcs1 plugin installed from charon

What do you mean by that? You obviously need a lot more plugins to generate and load keys and certificates. What's the plugin list in `pki --help`?

By the way, `pki --issue` can load private keys directly (`--type priv` or a specific type e.g. `--type rsa`). There is no need to use `pki --pub`.

**Reply by brobbinsharris** [reply]
I fixed my problem - my issue was that i was not using the correct ca_cert.pem

thanks for the help

**Reply by brobbinsharris** [reply]
Sorry missed your last reply - yes there are many other plugins installed - i just wanted to trim down the other pkcs's to try and isolate my issue.

thanks

**Reply by brobbinsharris** [reply]
Hi - I have a follow-up question.

Once our SA's for both flows are setup OK we are able to ping successfully.

We are unable to perform an ftpget or ftpput from our Linux client.

we are communicating between 2 nodes.
Both IP addressed are on the same subnet.

192.168.0.10 and 192.168.0.11.

We are setup for tunnel mode.

Any reason why ftpget and ftpput will not work through the tunnel?

thanks

**Reply by brobbinsharris** [reply]
our ftpget and ftpput work correctly from the Linux client when IPSEC is not running

**Reply by tobiasbrunner** [reply]
MTU issues could be a reason, but it seems weird in a local network (but since you use a very old version the kernel might be very old too, so anything is possible). Other than that, you have to find out where the traffic is dropped (check traffic counters, do traffic captures etc.).

**Reply by brobbinsharris** [reply]
thanks for the quick reply - i also see this issue when using strongswan version 5.8.1.

I simplified my test case to ftp a very small file of 1420 bytes and also 12 bytes and still the same issue.

without IPSEC running in strongswan - ftp transfers successfully without any issues.

on the ftpget from strongswan (linux) i see connection timed out

also tried both tunnel and transport modes

this is a point to point connection between 2 nodes 
192.168.0.10
192.168.0.11

both sides have MTU of 1500

thanks for the help - i will enable wireshark on the TFP server end to see what's going on with the ETH interface.

**Reply by brobbinsharris** [reply]
any issues with Port Numbers possibly?

**Reply by tobiasbrunner** [reply]
What do you mean?

**Reply by brobbinsharris** [reply]
just more of a general question with FTP and IPSEC ports if all protocols using different ports all processed correctly through the tunnel

**Reply by tobiasbrunner** [reply]
IPsec protects the complete IP packets, so the ports and protocols don't matter (unless you configured specific ports/protocols in the IPsec policies to only protect certain traffic).

**Reply by brobbinsharris** [reply]
Hi - I am looking to see the performance of different Diffie Hellman algorithms 4096 v 3072 and ecp384 and the time it takes to rekey the IKE SA and the IPSEC Child SA. Is there a debug switch I can turn on that shows the time it took to rekey in the resolution of milliseconds - so I can see the timing in syslog?

**Reply by tobiasbrunner** [reply]
There isn't. You can find [some numbers](https://wiki.strongswan.org/projects/strongswan/wiki/PublicKeySpeed) in our old wiki. If you want to test on a particular system with specific plugins you could use the `scripts/dh_speed` utility (it's built with `--enable-scripts`).

**Reply by brobbinsharris** [reply]
Thanks for the Info 
As I mentioned i am trying to analyze the performance of DH4096 v DH3072 and ecp384. My two Linux nodes have different versions of strongswan running.
nodeA = version 4.5.2
nodeB = version 5.8.1

I am unable to change these versions unfortunately.

When configuring my endpoints for rekeying i set "lifetime=5m" and "ikelifetime=5m". I do see the rekeying initiated but at a very faster rate in the order of 1 to 2 seconds.

I am unable to set the rekeying rate to exactly the time i specify - is there any issues with these older strongswan versions?

Once the rekeying initiates the entire SA rekey gets to the end when deleting the old SA. and I get these messages.

"deleting policy 192.168.0.11"
"Policy still used by another Child_SA, not removed"

can you help to explain why i am unable to delete the old SA?

thanks

**Reply by tobiasbrunner** [reply]
> When configuring my endpoints for rekeying i set "lifetime=5m" and "ikelifetime=5m". I do see the rekeying initiated but at a very faster rate in the order of 1 to 2 seconds.

With ipsec.conf, you have to change other settings besides these two, see [here](https://docs.strongswan.org/docs/5.9/config/rekeying.html) for the settings and formula.

> can you help to explain why i am unable to delete the old SA?

The messages are regarding policies, not the SA. Since the new SA after the rekeying uses the same policies they are not removed from the kernel.

**Reply by brobbinsharris** [reply]
Thank you. The current rekey parameters appear to only rekey the IPSEC CHILD SA. Is there a way to rekey both the IKE SA as well as the IPSEC SA?

**Reply by tobiasbrunner** [reply]
Please read the documentation.

**Reply by brobbinsharris** [reply]
Thank you for all the help over the course of this thread. I was able to rekey both the IKE and child IPSEC SA's and also included tcpdump into the rootfs - with this timestamp resolution I was able to deduce the performance of all the algorithms during the rekey phase between 2 Linux host embedded platforms.

thanks again.
